ocaml/xapi/import_raw_vdi.ml

```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
               debug "Checking whether localhost can see SR: %s" (Ref.string_of sr);
               if (Importexport.check_sr_availability ~__context sr)
               then localhost_handler rpc session_id vdi req s
-              else
-                let host = Importexport.find_host_for_sr ~__context sr in
-                let address = Db.Host.get_address ~__context ~self:host in
-                return_302_redirect req s address;
-                None
+              else 
+                begin
+                  try 
+                    let host = Importexport.find_host_for_sr ~__context sr in
+                    let address = Db.Host.get_address ~__context ~self:host in
+                    return_302_redirect req s address;
+                    None
+                  with 
+                  | Api_errors.Server_error (a,b) as e ->
+                    error "Caught exception in import handler: %s" (ExnHelper.string_of_exn e);
+                    let all = req.Request.cookie @ req.Request.query in 
+                    let task_id = 
+                    if List.mem_assoc "task_id" all 
+                    then Some (Ref.of_string (List.assoc "task_id" all))
+                    else None in
```

Comment:
I think this is a good opportunity to factor out some duplication :). If you check `Xapi_http.assert_credentials_ok` and `Xapi_http.with_context`, the code getting the `task_id` and `subtask_of` from the request is duplicated between them. I think these could be extracted into `task_id_of_req` and `subtask_of_of_req` helpers in `importexport.ml`.